# Step 1 - Read in and Inspect Data

In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.linear_model import LinearRegression

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.metrics import mean_squared_error

In [4]:
df1 = pd.read_csv('datasets/geo_data_0.csv')

In [5]:
df2 = pd.read_csv('datasets/geo_data_1.csv')

In [6]:
df3 = pd.read_csv('datasets/geo_data_2.csv')

In [7]:
df1

id        f0        f1        f2     product
0      txEyH  0.705745 -0.497823  1.221170  105.280062
1      2acmU  1.334711 -0.340164  4.365080   73.037750
2      409Wp  1.022732  0.151990  1.419926   85.265647
3      iJLyR -0.032172  0.139033  2.978566  168.620776
4      Xdl7t  1.988431  0.155413  4.751769  154.036647
...      ...       ...       ...       ...         ...
99995  DLsed  0.971957  0.370953  6.075346  110.744026
99996  QKivN  1.392429 -0.382606  1.273912  122.346843
99997  3rnvd  1.029585  0.018787 -1.348308   64.375443
99998  7kl59  0.998163 -0.528582  1.583869   74.040764
99999  1CWhH  1.764754 -0.266417  5.722849  149.633246

[100000 rows x 5 columns]

In [8]:
df2

id         f0         f1        f2     product
0      kBEdx -15.001348  -8.276000 -0.005876    3.179103
1      62mP7  14.272088  -3.475083  0.999183   26.953261
2      vyE1P   6.263187  -5.948386  5.001160  134.766305
3      KcrkZ -13.081196 -11.506057  4.999415  137.945408
4      AHL4O  12.702195  -8.147433  5.004363  134.766305
...      ...        ...        ...       ...         ...
99995  QywKC   9.535637  -6.878139  1.998296   53.906522
99996  ptvty -10.160631 -12.558096  5.005581  137.945408
99997  09gWa  -7.378891  -3.084104  4.998651  137.945408
99998  rqwUm   0.665714  -6.152593  1.000146   30.132364
99999  relB0  -3.426139  -7.794274 -0.003299    3.179103

[100000 rows x 5 columns]

In [9]:
df3

id        f0        f1        f2     product
0      fwXo0 -1.146987  0.963328 -0.828965   27.758673
1      WJtFt  0.262778  0.269839 -2.530187   56.069697
2      ovLUW  0.194587  0.289035 -5.586433   62.871910
3      q6cA6  2.236060 -0.553760  0.930038  114.572842
4      WPMUX -0.515993  1.716266  5.899011  149.600746
...      ...       ...       ...       ...         ...
99995  4GxBu -1.777037  1.125220  6.263374  172.327046
99996  YKFjq -1.261523 -0.894828  2.524545  138.748846
99997  tKPY3 -1.199934 -2.957637  5.219411  157.080080
99998  nmxp2 -2.419896  2.417221 -5.548444   51.795253
99999  V9kWn -2.551421 -2.025625  6.090891  102.775767

[100000 rows x 5 columns]

In [10]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [11]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [12]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [13]:
df1.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.343769      16.003790     185.364347

In [14]:
df2.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.734063       5.019721     137.945408

In [15]:
df3.describe()

f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.844801      16.739402     190.029838

In [16]:
df1.drop('id', axis=1, inplace=True)
df2.drop('id', axis=1, inplace=True)
df3.drop('id', axis=1, inplace=True)

While I would be more comfortable knowing more about the features, our initial data inspection shows no missing data. For a linear regression approach, we do not need the id of the well in question, so we drop that before proceeding. Since we know f0, f1, f2, and product are all important features, we will proceed with our analysis. 

# Step 2 - Training Linear Regression

In [17]:
X1 = df1.drop('product', axis=1)
y1 = df1['product']

In [18]:
X2 = df2.drop('product', axis=1)
y2 = df2['product']

In [19]:
X3 = df3.drop('product', axis=1)
y3 = df3['product']

In [20]:
X_train1, X_val1, y_train1, y_val1 = train_test_split(X1, y1, test_size=0.25, random_state=47)

In [21]:
X_train2, X_val2, y_train2, y_val2 = train_test_split(X2, y2, test_size=0.25, random_state=47)

In [22]:
X_train3, X_val3, y_train3, y_val3 = train_test_split(X3, y3, test_size=0.25, random_state=47)

In [23]:
model = LinearRegression()

In [24]:
model_1_score = cross_val_score(model, X1, y1, cv=5)

In [25]:
print('Average Model Score:', np.mean(model_1_score))

Average Model Score: 0.2754913072690447


In [26]:
model.fit(X_train1, y_train1)

LinearRegression()

In [27]:
predict_1 = model.predict(X_val1)

In [28]:
print('Average Predicted Output for Region 1:', np.mean(predict_1))

Average Predicted Output for Region 1: 92.360864667356


In [29]:
rmse1 = np.sqrt(mean_squared_error(y_val1, predict_1))

In [30]:
print('RMSE:', rmse1)

RMSE: 37.52379045981726


At this point, it looks like our linear regression has a higher RMSE. This suggests some large potential for variability in the wells. But, overall we do predict a return of 92,000 barrels in this region. 

In [31]:
model_2_score = cross_val_score(model, X2, y2, cv=5)

In [32]:
print('Average Model Score:', np.mean(model_2_score))

Average Model Score: 0.9996243728923553


In [33]:
model.fit(X_train2, y_train2)

LinearRegression()

In [34]:
predict_2 = model.predict(X_val2)

In [35]:
print('Average Predicted Output for Region 2:', np.mean(predict_2))

Average Predicted Output for Region 2: 68.81389806475146


In [36]:
rmse2 = np.sqrt(mean_squared_error(y_val2, predict_2))

In [37]:
print('RMSE:', rmse2)

RMSE: 0.8887286598305337


While the overall average in region 2 is lower than region one, we note here that the RMSE is much lower. This provides some more confidence that wells in this area should remain closer to the average prediction. Also, the extraordinarily high CV-score suggests that this model predicts the data very accurately and consistently. 

In [38]:
model_3_score = cross_val_score(model, X3, y3, cv=5)

In [39]:
print('Average Model Score:', np.mean(model_3_score))

Average Model Score: 0.1987156246205129


In [40]:
model.fit(X_train3, y_train3)

LinearRegression()

In [41]:
predict_3 = model.predict(X_val3)

In [42]:
print('Average Predicted Output for Region 3:', np.mean(predict_3))

Average Predicted Output for Region 3: 95.0912923012818


In [43]:
rmse3 = np.sqrt(mean_squared_error(y_val3, predict_3))

In [44]:
print('RMSE:', rmse3)

RMSE: 39.988392017930316


Finally, our third region suggests the highest predicted well output, but once again our RMSE is large. This adds in some variability. And, it also has the lowest CV-score, which suggests our data is less trust-worthy. 

# Step 3 - Prepare Data for Calculations

In [45]:
region_1_pred = pd.DataFrame(predict_1, index=X_val1.index, columns=['output'])

In [46]:
region_2_pred = pd.DataFrame(predict_2, index=X_val2.index, columns=['output'])

In [47]:
region_3_pred = pd.DataFrame(predict_3, index=X_val3.index, columns=['output'])

In [48]:
COST_PER_DEV = 100000000
PROFIT_PER_UNIT = 4500
WELLS_PER_DEV = 200
COST_PER_WELL = COST_PER_DEV / WELLS_PER_DEV
RISK = 0.025

In [49]:
minimum_units = COST_PER_DEV / PROFIT_PER_UNIT
minimum_per_well = minimum_units / 200
print('Minimum Yield (in Thousands) required for Profit:', minimum_units)
print('Minimum Yield required per well (in Thousands):', minimum_per_well)

Minimum Yield (in Thousands) required for Profit: 22222.222222222223
Minimum Yield required per well (in Thousands): 111.11111111111111


In [50]:
reg_1_top_200 = region_1_pred.sort_values(by='output', ascending=False)[:200]

In [51]:
mean_1 = reg_1_top_200.mean()
print(mean_1)

output    155.146466
dtype: float64


In [52]:
reg_2_top_200 = region_2_pred.sort_values(by='output', ascending=False)[:200]

In [53]:
mean_2 = reg_2_top_200.mean()
print(mean_2)

output    138.757188
dtype: float64


In [54]:
reg_3_top_200 = region_3_pred.sort_values(by='output', ascending=False)[:200]

In [55]:
mean_3 = reg_3_top_200.mean()
print(mean_3)

output    149.235162
dtype: float64


Here we got together some of the key variables, including the total cost per well development, which includes 200 wells. We also looked at the profit per thousand barrels, which currently sits at $4500. With this in mind, we can find that the minimum yield for a 200-well development is 22,222. Divided by the 200 wells in a development, this means each well needs to produce at least 111,000 barrels. 

Since we know 500 wells are surveyed and the best 200 of those 500 will be chosen, the next thing we did is look at the best possible output based on the scenarios. This involved taking the top 200 wells from our predictions in each region. Based on this, our best chance looks to be region 1, followed by 3, and finally 2. However, we note again that in our modeling, region 2 seems to be the most predictable -i.e. the least volatile. 

With all this in mind, we'll want to do some additional testing to see how our models stack up against our basic calculations.

# Step 4 - Calculate Potential Profits

In [56]:
def calculate_profit(data):
    '''Data should include the dataframe and column of interest'''
    total = np.sum(data) * PROFIT_PER_UNIT
    final = total - COST_PER_DEV
    return final

In [57]:
reg_1_total_profit = calculate_profit(reg_1_top_200['output'])

In [58]:
np.sum(reg_1_top_200['output'])

31029.293235817724

In [59]:
print('Total Expected Profit from Top 200 Wells in Region 1: {:.2f}'.format(reg_1_total_profit))

Total Expected Profit from Top 200 Wells in Region 1: 39631819.56


In [60]:
reg_2_total_profit = calculate_profit(reg_2_top_200['output'])

In [61]:
np.sum(reg_2_top_200['output'])

27751.437680029852

In [62]:
print('Total Expected Profit from Top 200 Wells in Region 1: {:.2f}'.format(reg_2_total_profit))

Total Expected Profit from Top 200 Wells in Region 1: 24881469.56


In [63]:
reg_3_total_profit = calculate_profit(reg_3_top_200['output'])

In [64]:
np.sum(reg_3_top_200['output'])

29847.032458871003

In [65]:
print('Total Expected Profit from Top 200 Wells in Region 1: {:.2f}'.format(reg_3_total_profit))

Total Expected Profit from Top 200 Wells in Region 1: 34311646.06


This is an initial calculation based on the top 200 predictions for each region. It's important to note that this would be an optimistict prediction, especially considering region's 1 and 3's volatility. 

Another, more simple measure would be to look at the mean, which we will do next. 

In [66]:
reg_1_mean_profit = (155.146466 * PROFIT_PER_UNIT) * WELLS_PER_DEV - COST_PER_DEV
print('Total Profit based On Average Output: {:.2f}'.format(reg_1_mean_profit))

Total Profit based On Average Output: 39631819.40


In [67]:
reg_2_mean_profit = (138.757188 * PROFIT_PER_UNIT) * WELLS_PER_DEV - COST_PER_DEV
print('Total Profit based On Average Output: {:.2f}'.format(reg_2_mean_profit))

Total Profit based On Average Output: 24881469.20


In [68]:
reg_3_mean_profit = (149.235162 * PROFIT_PER_UNIT) * WELLS_PER_DEV - COST_PER_DEV
print('Total Profit based On Average Output: {:.2f}'.format(reg_3_mean_profit))

Total Profit based On Average Output: 34311645.80


As one might expect, taking the mean of the top 200 data points and calculating potential profit gives a similiar picture to the more complex complication. We see only a slight variation from the above calculations. 

Now, we will examine this more closely by taking a profit range. To do this, we will look back at our model RMSE, which gives us an expected potential variance. With this added into our original calculation, we can get a better idea of how profits could vary. 

In [69]:
def low_var(data, rmse):
    for col in data:
        total = data - rmse
    return total

In [70]:
def high_var(data, rmse):
    for col in data:
        total = data + rmse
    return total

In [71]:
reg_1_top_200['low'] = low_var(reg_1_top_200['output'], rmse1)

In [72]:
reg_1_top_200['high'] = high_var(reg_1_top_200['output'], rmse1)

In [73]:
reg_1_profit_high = calculate_profit(reg_1_top_200['high'])

In [74]:
reg_1_profit_low = calculate_profit(reg_1_top_200['low'])

In [75]:
print('Region 1')
print('Potential Low: {:.2f}'.format(reg_1_profit_low)) 
print('Potential High: {:.2f}'.format(reg_1_profit_high)) 

Region 1
Potential Low: 5860408.15
Potential High: 73403230.98


In [76]:
reg_2_top_200['low'] = low_var(reg_2_top_200['output'], rmse2)

In [77]:
reg_2_top_200['high'] = high_var(reg_2_top_200['output'], rmse2)

In [78]:
reg_2_profit_high = calculate_profit(reg_2_top_200['high'])

In [79]:
reg_2_profit_low = calculate_profit(reg_2_top_200['low'])

In [80]:
print('Region 2')
print('Potential Low: {:.2f}'.format(reg_2_profit_low)) 
print('Potential High: {:.2f}'.format(reg_2_profit_high)) 

Region 2
Potential Low: 24081613.77
Potential High: 25681325.35


In [81]:
reg_3_top_200['low'] = low_var(reg_3_top_200['output'], rmse3)

In [82]:
reg_3_top_200['high'] = high_var(reg_3_top_200['output'], rmse3)

In [83]:
reg_3_profit_high = calculate_profit(reg_3_top_200['high'])

In [84]:
reg_3_profit_low = calculate_profit(reg_3_top_200['low'])

In [85]:
print('Region 3')
print('Potential Low: {:.2f}'.format(reg_3_profit_low)) 
print('Potential High: {:.2f}'.format(reg_3_profit_high)) 

Region 3
Potential Low: -1677906.75
Potential High: 70301198.88


Here we added two new columns to our top_200 dataframe. Specifically, low and high columns. These were based on the RMSE, which indicates differences between predictions and observations. In other words, taking highs and lows by adding and subtracting this value, respectively, we get a basic idea of what sort of volatility we can expect. As above, this is not meant to be a perfect indicator. However, taken as a whole, we can get a better idea of what we might expect in each region. 

From a conservative standpoint, region 2 seems to offer the most reliable profit potential. On our 100,000,000 investment, we can expect a total profit around 25,000,000. Alternatively, the potential rewards of region 1 are hard to ignore -with a potential profit of 73,000,000 based on high estimates -nearly three times higher than region 2. Region 3 offers a similiarly high upside potential, but the downside is marked with a loss. 

Considering everything, Region 2 definitely seems the safest bet for a solid 25% profit. But, in region 1 we could see anywhere between 5% and 73%, assuming things go according to our models. As such, I would recommend taking the risk and considering region 1. But, rather than stop here, it's important we perform some final calculations. 

# Step 5 - Digging into Risk and Profit Potential

To reviewer -sorry for sending incomplete, but I'm having some trouble with this last part. I did most of the calculations (I think), but clearly they aren't working out right. Or, I'm just confused about something. Can you please indicate if you see what I am doing wrong? And, if not anything particularly wrong, please let me know what I have left out. 

In [86]:
state = np.random.RandomState(47)

In [87]:
def profit(predict, y, count):
    probs_sorted = predict.sort_values(ascending=False)
    selected = y[probs_sorted.index][:count]
    profit = selected.sum() * PROFIT_PER_UNIT - COST_PER_DEV
    return profit

Is it that simple? This seems closer, but am I still doing something wrong here?

In [88]:
reg_1_pred = pd.Series(region_1_pred['output'])

In [89]:
values_1 = []
for i in range(1000):
    y_subsample = y_val1.sample(n=500, replace=True, random_state=state)
    pred_subsample = reg_1_pred[y_subsample.index]
    values_1.append(profit(pred_subsample,
                          y_subsample,
                          200))

In [90]:
values_1 = pd.Series(values_1)

In [91]:
lower_1 = values_1.quantile(0.025)
upper_1 = values_1.quantile(0.975)
 
mean_1 = values_1.mean()
print("Average revenue in Region 1:", mean_1)
print("2.5% quantile in Region 1:", lower_1)
print("97.5% quantile in Region 1:", upper_1)

Average revenue in Region 1: 4872050.680696629
2.5% quantile in Region 1: -537540.9981569666
97.5% quantile in Region 1: 10091493.365440719


In [92]:
losses_1 = values_1[values_1 < 0]
len(losses_1)

36

In [93]:
reg_2_pred = pd.Series(region_2_pred['output'])

In [94]:
values_2 = []
for i in range(1000):
    y_subsample = y_val2.sample(n=500, replace=True, random_state=state)
    pred_subsample = reg_2_pred[y_subsample.index]
    values_2.append(profit(pred_subsample,
                          y_subsample,
                          200))

In [95]:
values_2 = pd.Series(values_2)

In [96]:
lower_2 = values_2.quantile(0.025)
upper_2 = values_2.quantile(0.975)
 
mean_2 = values_2.mean()
print("Average revenue in Region 2:", mean_2)
print("2.5% quantile in Region 2:", lower_2)
print("97.5% quantile in Region 2:", upper_2)

Average revenue in Region 2: 5069009.333665242
2.5% quantile in Region 2: 617217.9955411233
97.5% quantile in Region 2: 9222658.211552847


In [97]:
losses_2 = values_2[values_2 < 0]
len(losses_2)

9

In [98]:
reg_3_pred = pd.Series(region_3_pred['output'])

In [99]:
values_3 = []
for i in range(1000):
    y_subsample = y_val3.sample(n=500, replace=True, random_state=state)
    pred_subsample = reg_3_pred[y_subsample.index]
    values_3.append(profit(pred_subsample,
                          y_subsample,
                          200))

In [100]:
values_3 = pd.Series(values_3)

In [101]:
lower_3 = values_3.quantile(0.025)
upper_3 = values_3.quantile(0.975)
 
mean_3 = values_3.mean()
print("Average revenue in Region 3:", mean_3)
print("2.5% quantile in Region 3:", lower_3)
print("97.5% quantile in Region 3:", upper_3)

Average revenue in Region 3: 4146265.4446982797
2.5% quantile in Region 3: -1668471.5449371152
97.5% quantile in Region 3: 9527435.472567124


In [102]:
losses_3 = values_3[values_3 < 0]
len(losses_3)

76

# Recommendation

Starting with Region 3, we see fairly high profit potential, but the average is not impressive. Further, there is the largest potential for loss in this region, which should immediately exclude it from the running. Region 1 offers the highest profit potential, but maintains some volatility. In fact, we see 4 times as many potential losses in our bootstrapped sample than we do in region 2.

Specifically, in Region 2 we se even the lowest 2.5% of our samples would still pull with a small profit. Further, it maintaines the highest average profit with our sampling. With all this in mind, I recommend going with the safer alternative, which should all but guarantee profit. I recommend exploring Region 2 for development. 